Storage/storage.h

```diff
@@ -73,22 +73,47 @@ namespace storage {
+    using T_KEY = current::decay<decltype(::current::storage::sfinae::GetKey(std::declval<base>()))>; \
+    T_KEY key() const { return ::current::storage::sfinae::GetKey(data); }                            \
+    void set_key(const T_KEY& key) { ::current::storage::sfinae::SetKey(data, key); }                 \
+  };                                                                                                  \
+  CURRENT_STRUCT(alias##Deleted) {                                                                    \
+    CURRENT_FIELD(key, ::current::storage::sfinae::ENTRY_KEY_TYPE<base>);                             \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##Deleted) {}                                                    \
+    CURRENT_CONSTRUCTOR(alias##Deleted)(const base& value)                                            \
+        : key(::current::storage::sfinae::GetKey(value)) {}                                           \
+  };                                                                                                  \
+  struct alias {                                                                                      \
+    using T_ENTRY = base;                                                                             \
+    using T_UPDATE_EVENT = alias##Updated;                                                            \
+    using T_DELETE_EVENT = alias##Deleted;                                                            \
+    using T_PERSISTED_EVENT_1 = alias##Updated;                                                       \
+    using T_PERSISTED_EVENT_2 = alias##Deleted;                                                       \
   }
+
+#define CURRENT_STRUCT_TAG_Vector(base, alias)                                    \
+  CURRENT_STRUCT(alias##PushedElement) {                                          \
+    CURRENT_FIELD(data, base);                                                    \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##PushedElement) {}                          \
+    CURRENT_CONSTRUCTOR(alias##PushedElement)(const base& value) : data(value) {} \
+  };                                                                              \
+  CURRENT_STRUCT(alias##PoppedElement) {};                                        \
+  struct alias {                                                                  \
+    using T_ENTRY = base;                                                         \
+    using T_PUSHED_BACK_EVENT = alias##PushedElement;                             \
+    using T_POPPED_BACK_EVENT = alias##PoppedElement;                             \
+    using T_PERSISTED_EVENT_1 = alias##PushedElement;                             \
+    using T_PERSISTED_EVENT_2 = alias##PoppedElement;                             \
+  }
+
+// TODO(dkorolev) || TODO(mzhurovich): Move to `TypeSystem/` and test.
+#define CURRENT_STRUCT_TAG(tag, base, alias) CURRENT_STRUCT_TAG_##tag(base, alias)
```

Comment:
And this semantics became much less obvious for the user :(
Looking just at the name of the macro `CURRENT_STRUCT_TAG` I suppose it to implement some way to "tag" any Current structure. So I think about possible `MyStruct`<->`SomeTag` connection.
Surprise! What this macro does has nothing in common with my expectations :)
* `tag` means storage field container type
* `alias` actually means `tag`
* this whole thing is supposed to be used for Storage. WTF?! :confounded: